### PR TITLE
feat: allow root path to run the app on `/path`

### PR DESCRIPTION
**Pull request by pritamsoni-hsr**
## What does this PR do?

allow apps to be served from a custom path when running app behind a proxy.

by default apps are served from `https://<app_id>.litng-ai-03.litng.ai`, this PR adds a feature so apps can be served from 'https://mydomain.com/<my_app>'.

if you want to build custom UI, make sure all internal links are prefixed with `window.app_prefix`

- this works only behind a proxy and not on the `api_id.litng-ai-03.litng.ai/<custom_path>` and this is because of uvicorn's root path param.

## Limitations 
relative links inside the app will break, so user should use `window.app_prefix` in their JS, there is no warning around that atm.

## breaking changes
none

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by hhsecond** [comment]
> relative links inside the app will break, so user should use window.app_prefix in their JS, there is no warning around that atm.
Can this be updated in the docs?

File: src/lightning_app/core/app.py [hunk 1]
```diff
@@ -67,6 +68,9 @@ def __init__(
                 This can be helpful when reporting bugs on Lightning repo.
             info: Provide additional info about the app which will be used to update html title,
                 description and image meta tags and specify any additional tags as list of html strings.
+            root_path: Set this to `/path` if you want to run your app behind a proxy at `/path` leave empty for "/".
```
Comment (hhsecond) [inline src/lightning_app/core/app.py:71]: can you clarify what's proxy in this context? Our users might not know what would that be? Perhaps an example?

Also, in the next line, it is mentioned that the app won't start but not clear what should be done for it to actually work!

File: src/lightning_app/utilities/frontend.py [hunk 2]
```diff
@@ -27,19 +32,29 @@ def update_index_file_with_info(ui_root: str, info: AppInfo = None) -> None:
...
-    if not info:
-        return
+    if info:
+        with original_file.open() as f:
+            original = f.read()
 
-    original = ""
+        with entry_file.open("w") as f:
+            f.write(
+                rewrite_static_with_root_path(_get_updated_content(original=original, root_path=root_path, info=info))
```
Comment (hhsecond) [inline src/lightning_app/utilities/frontend.py:41]: do we have to pipe this call? Can we `rewrite_static_with_root_path` happens inside `_get_updated_content`.

And probably not even a function, just a section inside `_get_updated_content`
?

File: src/lightning_app/utilities/frontend.py [hunk 3]
```diff
@@ -14,10 +14,15 @@ class AppInfo:
     meta_tags: Optional[List[str]] = None
 
 
-def update_index_file_with_info(ui_root: str, info: AppInfo = None) -> None:
+def update_index_file_with_info_and_root_path(
```
Comment (hhsecond) [inline src/lightning_app/utilities/frontend.py:17]: perhaps just `update_index_file`?

Comment (hhsecond) [reply src/lightning_app/utilities/frontend.py:17]: the arguments says what's being used for updating it

**Review by hhsecond — COMMENTED** [review]
Some minor nits and clarification on the docs

**Review by tchaton — APPROVED** [review]
LGTM!
@pritamsoni-hsr Do you think it would make sense to extract those metadata directly into a private file under a `.lightning` folder?

**Comment by Borda** [comment]
> Do you think it would make sense to extract those metadata directly into a private file under a `.lightning` folder?

let's leave it s separate feature :)

File: src/lightning_app/utilities/frontend.py (hunk 3 above)
Comment (kaushikb11) [reply src/lightning_app/utilities/frontend.py:17]: good point!

Comment (kaushikb11) [reply src/lightning_app/utilities/frontend.py]: updated!

File: src/lightning_app/core/app.py [hunk 4]
```diff
@@ -49,6 +49,7 @@ def __init__(
         root: "lightning_app.LightningFlow",
         debug: bool = False,
         info: frontend.AppInfo = None,
+        root_path: str = "",
```
Comment (awaelchli) [inline src/lightning_app/core/app.py:52]: the text below says that for customization, I should set it to "/path". By extension, shouldn't this then default to "/"?

File: src/lightning_app/core/app.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning_app/core/app.py:71]: I agree that the second sentence should probably point to some resource for further details :)

File: src/lightning_app/frontend/web.py [hunk 5]
```diff
@@ -20,6 +20,7 @@ class StaticWebFrontend(Frontend):
 
     Arguments:
         serve_dir: A local directory to serve files from. This directory should at least contain a file `index.html`.
+                root_path: A path prefix when routing traffic from behind a proxy at `<root_path>`
```
Comment (awaelchli) [inline src/lightning_app/frontend/web.py:23]: malformated

File: src/lightning_app/frontend/frontend.py [hunk 6]
```diff
@@ -15,21 +15,21 @@ def __init__(self) -> None:
...
+    def start_server(self, host: str, port: int, root_path: str = "") -> None:
         """Start the process that serves the UI at the given hostname and port number.
 
         Arguments:
             host: The hostname where the UI will be served. This gets determined by the dispatcher (e.g., cloud),
                 but defaults to localhost when running locally.
             port: The port number where the UI will be served. This gets determined by the dispatcher, which by default
                 chooses any free port when running locally.
-
+            root_path: root_path for the server if app in exposed via a proxy at `/<root_path>`
```
Comment (awaelchli) [inline src/lightning_app/frontend/frontend.py:26]:
```suggestion
            root_path: root_path for the server if app in exposed via a proxy at `/<root_path>`
            
```
You need to leave an empty line

Comment (awaelchli) [reply src/lightning_app/frontend/frontend.py:26]: Can we include an example like `https://domain.com/<root_path>/my_page` to visualize what it means.

File: tests/tests_app/frontend/test_web.py [hunk 7]
```diff
@@ -54,17 +55,21 @@ def test_start_server_through_function(uvicorn_mock, tmpdir, monkeypatch):
     FastAPIMock.get.return_value = FastAPIGetDecoratorMock
     monkeypatch.setattr(lightning_app.frontend.web, "FastAPI", MagicMock(return_value=FastAPIMock))
 
-    lightning_app.frontend.web.start_server(serve_dir=tmpdir, host="myhost", port=1000, path="/test-flow")
-    uvicorn_mock.run.assert_called_once_with(app=ANY, host="myhost", port=1000, log_config=ANY)
-    FastAPIMock.mount.assert_called_once_with("/test-flow", ANY, name="static")
+    root_path = "/base"
```
Comment (awaelchli) [inline tests/tests_app/frontend/test_web.py:58]: could you parametrize the test with `root_path=("/", "/base")` please?

File: src/lightning_app/utilities/frontend.py (hunk 2 above)
Comment (kaushikb11) [reply src/lightning_app/utilities/frontend.py]: updated!

File: src/lightning_app/core/app.py (hunk 4 above)
Comment (kaushikb11) [reply src/lightning_app/core/app.py:52]: uvicorn and fastapi both have empty string as default:
https://github.com/encode/uvicorn/blob/master/uvicorn/main.py#L248
https://github.com/tiangolo/fastapi/blob/master/fastapi/applications.py#L77

File: tests/tests_app/frontend/test_web.py (hunk 7 above)
Comment (kaushikb11) [reply tests/tests_app/frontend/test_web.py]: updated!

**Review by ethanwharris — APPROVED** [review]

File: src/lightning_app/core/app.py (hunk 1 above)
Comment (kaushikb11) [reply src/lightning_app/core/app.py:71]: @pritamsoni-hsr

**Review by manskx — APPROVED** [review]

File: src/lightning_app/core/app.py (hunk 1 above)
Comment (pritamsoni-hsr) [reply src/lightning_app/core/app.py:71]: can we add this url? https://www.fortinet.com/resources/cyberglossary/proxy-server

**Comment by pritamsoni-hsr** [comment]
@tchaton we could, but I think maintaining it in the app is easiest, we can later serialize and deserialize `.lightning` file

**Review by kaushikb11 — APPROVED** [review]

**Pull request merged**